include/boost/geometry/util/type_traits.hpp

```diff
@@ -0,0 +1,323 @@
+// Boost.Geometry
+
+// Copyright (c) 2020, Oracle and/or its affiliates.
+
+// Contributed and/or modified by Adam Wulkiewicz, on behalf of Oracle
+
+// Licensed under the Boost Software License version 1.0.
+// http://www.boost.org/users/license.html
+
+#ifndef BOOST_GEOMETRY_UTIL_TYPE_TRAITS_HPP
+#define BOOST_GEOMETRY_UTIL_TYPE_TRAITS_HPP
+
+
+#include <iterator>
+#include <type_traits>
+
+#include <boost/geometry/core/tag.hpp>
+#include <boost/geometry/core/tags.hpp>
+
+#include <boost/mpl/has_xxx.hpp>
+#include <boost/range/has_range_iterator.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail
+{
+
+
+// C++17
+template <bool B>
+using bool_constant = std::integral_constant<bool, B>;
+
+
+// C++17
+template <typename ...>
+struct conjunction
+    : std::true_type
+{};
+template<typename Trait>
+struct conjunction<Trait>
+    : Trait
+{};
+template <typename Trait, typename ...Traits>
+struct conjunction<Trait, Traits...>
+    : std::conditional_t<Trait::value, conjunction<Traits...>, Trait>
+{};
+
+
+// C++17
+template <typename ...>
+struct disjunction
+    : std::false_type
+{};
+template <typename Trait>
+struct disjunction<Trait>
+    : Trait
+{};
+template <typename Trait, typename ...Traits>
+struct disjunction<Trait, Traits...>
+    : std::conditional_t<Trait::value, Trait, disjunction<Traits...>>
+{};
+
+
+// C++17
+template <typename Trait>
+struct negation
+    : bool_constant<!Trait::value>
+{};
+
+
+/*
+template <typename ...Traits>
+using and_ = conjunction<Traits...>;
+
+template <typename ...Traits>
+using or_ = disjunction<Traits...>;
+
+template <typename Trait>
+using not_ = negation<Trait>;
+*/
+
+
+template <typename T>
+struct is_geometry
+    : bool_constant<! std::is_void<typename tag<T>::type>::value>
+{};
+
+template <typename T>
+struct is_not_geometry
+    : std::is_void<typename tag<T>::type>
+{};
+
+
+template <typename T>
+struct is_point
+    : std::is_same<point_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_multi_point
+    : std::is_same<multi_point_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_pointlike
+    : std::is_base_of<pointlike_tag, typename tag<T>::type>
+{};
+
+
+template <typename T>
+struct is_segment
+    : std::is_same<segment_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_linestring
+    : std::is_same<linestring_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_multi_linestring
+    : std::is_same<multi_linestring_tag, typename tag<T>::type>
+{};
+
+template <typename T>
+struct is_polylinear
+    : std::is_base_of<polylinear_tag, typename tag<T>::type>
+{};
```

Comment:
I decided to use *polylinear* name for linear geometries which are not single segment. Similar to *polygonal* for areal geometries that are not a box. If you have a different idea feel free to propose a change.

Reply:
What is the difference with `linestring` ? Those are also linear geometries with not a single segment. But I assume you mean something else?

Reply:
Yes, polylinear means linestring or multi_linestring. I added the tag in case someone wanted to use `tag_cast` instead and because we already have `polygonal_tag`.